src/c/c_client.cpp

```diff
@@ -30,517 +30,654 @@
+  }
+  
+  std::string key_str(key, key_length);
 
   std::vector<std::string> input_vec;
-  for(size_t i=0; i<n_inputs; i++)
-    input_vec.push_back(std::string(inputs[i], input_lengths[i]));
+  if (1 != n_inputs || 0 != input_lengths[0]) {
+    for (size_t i=0; i<n_inputs; i++) {
+      input_vec.push_back((NULL != inputs[i] && 0 != input_lengths[i]) ?
+			  std::string(inputs[i], input_lengths[i]) :
+			  std::string(""));
+    }
+  }
 
   std::vector<std::string> output_vec;
-  for(size_t i=0; i<n_outputs; i++)
-    output_vec.push_back(std::string(outputs[i], output_lengths[i]));
+  if (1 != n_outputs || 0 != output_lengths[0]) {
+    for (size_t i=0; i<n_outputs; i++) {
+      output_vec.push_back((NULL != outputs[i] && 0 != output_lengths[i]) ?
+			   std::string(outputs[i], output_lengths[i]) :
+			   std::string(""));
+    }
+  }
 
+  Client *s = reinterpret_cast<Client *>(c_client);
   s->run_model(key_str, input_vec, output_vec);
-  return;
 }
 
+// Check whether a key exists in the database
 extern "C"
-bool key_exists(void* c_client, const char* key,
-                const size_t key_length)
+bool key_exists(void *c_client, const char *key, const size_t key_length)
 {
-  Client* s = (Client *)c_client;
-  std::string key_str = std::string(key, key_length);
+  // Sanity check params
+  if (NULL == c_client || NULL == key)
+    return false;
+
+  Client *s = reinterpret_cast<Client *>(c_client);
+  std::string key_str(key, key_length);
+
   return s->key_exists(key_str);
 }
 
+// Check whether a model exists in the database
 extern "C"
-bool model_exists(void* c_client, const char* name,
-                  const size_t name_length)
+bool model_exists(void *c_client, const char *name, const size_t name_length)
 {
-  Client* s = (Client *)c_client;
-  std::string name_str = std::string(name, name_length);
+  // Sanity check params
+  if (NULL == c_client || NULL == name)
+    return false;
+
+  Client *s = reinterpret_cast<Client *>(c_client);
+  std::string name_str(name, name_length);
+
   return s->model_exists(name_str);
 }
 
+// Check whether a tensor exists in the database
 extern "C"
-bool tensor_exists(void* c_client, const char* name,
-                   const size_t name_length)
+bool tensor_exists(void *c_client, const char *name, const size_t name_length)
 {
-  Client* s = (Client *)c_client;
-  std::string name_str = std::string(name, name_length);
+  // Sanity check params
+  if (NULL == c_client || NULL == name)
+    return false;
+
+  Client *s = reinterpret_cast<Client *>(c_client);
+  std::string name_str(name, name_length);
+
   return s->tensor_exists(name_str);
 }
 
+// Delay until a key exists in the database
 extern "C"
-bool poll_key(void* c_client,
-              const char* key, const size_t key_length,
-              const int poll_frequency_ms, const int num_tries)
+bool poll_key(void *c_client,
+              const char *key,
+	      const size_t key_length,
```

Comment:
Check that types are aligned vertically when possible.

Reply:
probably a tab/space issue from my editor. Will fix.